src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,246 @@
+    /// <param name="image">The decoded image.</param>
+    /// <returns><see langword="true"/> if the image should be resized, otherwise; <see langword="false"/>.</returns>
+    private static bool ShouldResize(DecoderOptions options, Image image)
+    {
+        if (options.TargetSize is null)
+        {
+            return false;
+        }
+
+        Size targetSize = options.TargetSize.Value;
+        Size currentSize = image.Size();
+        return currentSize.Width != targetSize.Width && currentSize.Height != targetSize.Height;
+    }
+
+    internal static T WithSeekableStream<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, T> action)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        T Action(Stream s, long position)
+        {
+            T result = action(s);
+
+            // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
+
+            return result;
+        }
+
+        if (stream.CanSeek)
+        {
+            return Action(stream, stream.Position);
+        }
+
+        Configuration configuration = options.Configuration;
+        using ChunkedMemoryStream memoryStream = new(configuration.MemoryAllocator);
+        stream.CopyTo(memoryStream, configuration.StreamProcessingBufferSize);
+        memoryStream.Position = 0;
+
+        return Action(memoryStream, 0);
+    }
+
+    internal static async Task<T> WithSeekableStreamAsync<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, CancellationToken, T> action,
+        CancellationToken cancellationToken)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        T Action(Stream s, long position, CancellationToken ct)
+        {
+            T result = action(s, ct);
+
+            // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
+
+            if (ct.IsCancellationRequested)
+            {
+                throw new TaskCanceledException();
+            }
```

Comment:
I don't see value in this check after we effectively finished the decoding.

Reply:
I'll see if our tests pass if I move it before. Basically, none of our decoders are currently checking for this and I didn't want to have to inject it into each one.

Reply:
Jpeg and tiff do:
https://github.com/SixLabors/ImageSharp/search?q=ThrowIfCancellationRequested

It's good to check for cancellation before calling the decoders, I see no point to do it afterwards.

Reply:
OK. Let me double check the tests. Maybe we're pushing to a decoder that doesn't.

Reply:
The [general tests](https://github.com/SixLabors/ImageSharp/blob/5834c39c41a4bfeaef5d02b43d1d818e3c5c467b/tests/ImageSharp.Tests/Image/ImageTests.Decode_Cancellation.cs) here are not great. They don't test any of the decoders rather a mocked infrastructure that doesn't contain any viable image data (the header is a `Guid`). This means that the checks have to be in `WithSeekableStream` in order for those tests to pass.

IMO this highlights the dangers of using mocks. I would remove every single instance of usage if I could find the time. 

It is possible to use `BufferedReadStream` btw, you can pass the cancellation token via the constructor.

Reply:
OK. I've replaced those tests with ones that test `IdentifyAsync(...)` and `DecodeAsync(...)` for all our decoders.

`BufferedReadStream .Read(Span<byte> buffer)` now throws if the token is cancelled.